tools/qemu-kvm-patch/qemu-2.8-memaccess.patch

```diff
@@ -0,0 +1,362 @@
++    if (pargs == NULL){
++        error_setg(errp, "Qemu pmemaccess: malloc failed");
++        return;
++    }
++
++    pargs->errp = errp;
++    // create a copy of path that we can safely use
++    size_t path_size = strlen(path);
++    pargs->path = malloc(path_size + 1);
++    memcpy(pargs->path, path, path_size);
++    pargs->path[path_size] = '\0';
++
++    // create socket
++    pargs->socket_fd = socket(PF_UNIX, SOCK_STREAM, 0);
++    if (pargs->socket_fd < 0){
++        error_setg(pargs->errp, "Qemu pmemaccess: socket failed");
++        return;
++    }
++    // unlink path if already exists
++    unlink(path);
++    // bind socket
++    pargs->address = malloc(sizeof(struct sockaddr_un));
++    if (pargs->address == NULL){
++        error_setg(pargs->errp, "Qemu pmemaccess: malloc failed");
++        return;
++    }
++    pargs->address->sun_family = AF_UNIX;
++    pargs->address_length = sizeof(pargs->address->sun_family) + sprintf(pargs->address->sun_path, "%s", (char *) pargs->path);
++    if (bind(pargs->socket_fd, (struct sockaddr *) pargs->address, pargs->address_length) != 0){
++        printf("could not bind\n");
++        error_setg(pargs->errp, "Qemu pmemaccess: bind failed");
++        return;
++    }
++
++    // listen
++    if (listen(pargs->socket_fd, 0) != 0){
++        error_setg(pargs->errp, "Qemu pmemaccess: listen failed");
++        return;
++    }
++
++    // start the thread
++    sigfillset(&set);
++    pthread_sigmask(SIG_SETMASK, &set, &oldset);
++    pthread_create(&thread, NULL, memory_access_thread, pargs);
++    pthread_sigmask(SIG_SETMASK, &oldset, NULL);
++}
+diff --git a/memory-access.h b/memory-access.h
+new file mode 100644
+index 0000000..21c7137
+--- /dev/null
++++ b/memory-access.h
+@@ -0,0 +1,24 @@
++/*
++ * Open A UNIX Socket access to physical memory
++ *
++ * Author: Valerio G. Aimale <valerio@aimale.com>
++ */
++
++#ifndef MEMORY_ACCESS_H
++#define MEMORY_ACCESS_H
++
++#include <sys/socket.h>
++#include "qemu/osdep.h"
++#include "qapi/error.h"
++
++void qmp_pmemaccess (const char *path, Error **errp);
++
++struct pmemaccess_args {
++    int socket_fd;
++    struct sockaddr_un *address;
++    socklen_t address_length;
++    char *path;
++    Error **errp;
++};
++
++#endif /* MEMORY_ACCESS_H */
+diff --git a/qapi-schema.json b/qapi-schema.json
+index a0d3b5d..9292ded 100644
+--- a/qapi-schema.json
++++ b/qapi-schema.json
+@@ -1679,6 +1679,34 @@
+   'data': {'val': 'int', 'size': 'int', 'filename': 'str'} }
+ 
+ ##
++# @pmemaccess:
++#
++# Open A UNIX Socket access to physical memory
++#
++# @path: the name of the UNIX socket pipe
++#
++# Returns: Nothing on success
++#
++# Since: 2.4.0.1
++#
++# Notes: Derived from previously existing patches. When command
++# succeeds connect to the open socket. Write a binary structure to
++# the socket as:
++#
++# struct request {
++#     uint8_t type;   // 0 quit, 1 read, 2 write, ... rest reserved
```

Comment:
So since this is an ABI I think it would be much better if we changed this to a uint64_t as well as to avoid various compilers deciding on different padding after it. I realize we have been using this in all other patches too but at least going forward we should be safer.